File: scripts/ping-indexnow.py
import urllib.request
import json
import xml.etree.ElementTree as ET

HOST = 'sadmanportfolio.vercel.app'
KEY = '9f31a293b8e1467cb40d5138f293bca1'
KEY_LOCATION = f'https://{HOST}/{KEY}.txt'

tree = ET.parse('public/sitemap.xml')
root = tree.getroot()
urls = [elem.text for elem in root.findall('{http://www.sitemaps.org/schemas/sitemap/0.9}url/{http://www.sitemaps.org/schemas/sitemap/0.9}loc') if elem.text]

payload = {
    'host': HOST,
    'key': KEY,
    'keyLocation': KEY_LOCATION,
    'urlList': urls
}

data = json.dumps(payload).encode('utf-8')
req = urllib.request.Request(
    'https://api.indexnow.org/IndexNow',
    data=data,
    headers={'Content-Type': 'application/json; charset=utf-8'}
)

try:
    with urllib.request.urlopen(req) as resp:
        print(f'IndexNow ping response: {resp.status} {resp.reason}')
        print(f'Successfully submitted {len(urls)} URLs to IndexNow (Bing/Copilot/Yandex)!')
except urllib.error.HTTPError as e:
        print(f'IndexNow HTTP response: {e.code} - {e.read().decode("utf-8", errors="ignore")}')
except Exception as e:
        print(f'IndexNow error: {e}')

File: scripts/ping-indexnow.ts
import { readFileSync } from 'fs';
import { resolve } from 'path';

const HOST = 'sadmanportfolio.vercel.app';
const KEY = '9f31a293b8e1467cb40d5138f293bca1';
const KEY_LOCATION = https://System.Management.Automation.Internal.Host.InternalHost/.txt;

async function pingIndexNow() {
  const xml = readFileSync(resolve('public/sitemap.xml'), 'utf-8');
  const urls: string[] = [];
  const regex = /<loc>(https?:\/\/[^<]+)<\/loc>/g;
  let match;
  while ((match = regex.exec(xml)) !== null) {
    urls.push(match[1]);
  }

  const payload = {
    host: HOST,
    key: KEY,
    keyLocation: KEY_LOCATION,
    urlList: urls,
  };

  try {
    const res = await fetch('https://api.indexnow.org/IndexNow', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json; charset=utf-8' },
      body: JSON.stringify(payload),
    });
    console.log([IndexNow] Status:  );
    console.log([IndexNow] Successfully pushed  URLs to Bing & IndexNow engines.);
  } catch (err) {
    console.error('[IndexNow] Error pinging IndexNow:', err);
  }
}

pingIndexNow();
